# Laboratorio de regresión - 3

## Significancia de factores

|                |                        |
:----------------|------------------------|
| **Nombre**     Camila Garcia Ornelas|   |
| **Fecha**     25 de agosto          |   |
| **Expediente 749725               * |   |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [56]:
import pandas as pd
import numpy as np

In [58]:
data = pd.read_csv("Advertising.csv")

In [60]:
data.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

Sí. Cuando se gasta más en publicidad, normalmente las ventas suben.

La publicidad hace que más personas conozcan el producto y lo recuerden; así es más probable que compren.

Si no hubiera relación, el gasto en publicidad no explicaría las ventas ni serviría para predecirlas. En ese caso buscaríamos otras variables (precio, disponibilidad, etc.) y no usaríamos el gasto publicitario para pronosticar.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

Podemos medir que tan fuerte es la relación dependiendo del R², el β₁ y su p‑valor.

Si r es cerca de 1 es fuerte; si r es cerca de 0 es débil. 

Si β₁ es positiva y su p‑valor < 0.05, hay evidencia de que la publicidad sí afecta a las ventas.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [63]:
from sklearn.linear_model import LinearRegression

In [116]:
x = data["TV"].values.reshape(-1, 1)
y = data["sales"].values.reshape(-1, 1)

In [93]:
lr = LinearRegression().fit(x, y)
print("β0:", modelo.intercept_)
print("β1", modelo.coef_[0])

β0: 7.032593549127695
β1 0.04753664043301975


### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [110]:
y_pred = lr.predict(x)
rss = np.sum((y - y_pred)**2)
n = len(y)
p = 2
sigma2 = rss / (n - p)
sigma2

10.618841328946218

In [118]:
x_mean = np.mean(x)
SE_b1 = np.sqrt(sigma2 / np.sum((x - x_mean)**2))
SE_b1

0.0026906071877968707

In [127]:
SE_b0 = np.sqrt(sigma2 * (1/n + x_mean**2 / np.sum((x - x_mean)**2)))
SE_b0

0.4578429402734785

Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [153]:
b0 = lr.intercept_
b1 = lr.coef_[0]
b0, b1

(7.032593549127695, 0.04753664043301975)

In [133]:
IC_b1 = (b1 - 2*SE_b1, b1 + 2*SE_b1)
IC_b1

(0.04215542605742601, 0.05291785480861349)

In [137]:
IC_b0 = (b0 - 2*SE_b0, b0 + 2*SE_b0)
IC_b0

(6.116907668580738, 7.948279429674652)

Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_0: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

La hipótesis nula quiere decir que los datos no traen evidencia para sugerir que hay un cambio. 

La hipótesis alternativa quiere decir que si hay evidencia, y es lo que queremos detectar.

H₀ es el “no hay efecto”; H₁ es el “sí hay efecto”. Rechazar H₀ no prueba que H₁ sea 100% verdadera, solo indica que los datos dan evidencia suficiente a su favor con el criterio elegido

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

In [147]:
t0 = (b0 - 0)/SE_b0

15.360275174117547

In [149]:
t1 = (b1-0)/SE_b1

17.667625600875546

In [151]:
[t0,t1]

[15.360275174117547, 17.667625600875546]

La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [155]:
from scipy import stats

In [161]:
p_b0 = 2*(1 - stats.t.cdf(np.abs(t0), n-p))
p_b0

0.0

In [163]:
p_b1 = 2*(1 - stats.t.cdf(np.abs(t1), n-p))
p_b1

0.0

¿Se rechaza la hipótesis nula? ¿Qué significa?

Se rechaza la hipotesis nula. Hay evidencia de que β₁ ≠ 0.

Esto quiere decir que la inversión en TV sí tiene efecto en las ventas.

Si no se rechazara la hipotesis, querría decir que la inversión en TV no tendría ningun efecto en las ventas. 

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

In [308]:
print("Regresión de periódico")
x2 = data["newspaper"].values.reshape(-1, 1)
y2 = data["sales"].values.reshape(-1, 1)

#coeficientes
lr2 = LinearRegression().fit(x2, y2)
print("β0:", lr2.intercept_)
print("β1:", lr2.coef_[0])

#error estándar de coeficientes
y_pred2 = lr2.predict(x2)
rss2 = np.sum((y2 - y_pred2)**2)
n2 = len(y2)
p2 = 2
sigma2_2 = rss2 / (n2 - p2)

x2_mean = np.mean(x2)
SE2_b1 = np.sqrt(sigma2_2 / np.sum((x2 - x2_mean)**2))
print(SE2_b1)

SE2_b0 = np.sqrt(sigma2_2 * (1/n2 + x2_mean**2 / np.sum((x2 - x2_mean)**2)))
print(SE2_b0)

#intervalos de confianza
b0_2 = lr2.intercept_
b1_2 = lr2.coef_[0]
print(b0_2, b1_2)

IC_b1_2 = (b1_2 - 2*SE2_b1, b1_2 + 2*SE2_b1)
print(IC_b1_2)

IC_b0_2 = (b0_2 - 2*SE2_b0, b0_2 + 2*SE2_b0)
print(IC_b0_2)

t0_2 = (b0_2 - 0)/SE2_b0
t1_2 = (b1_2-0)/SE2_b1

print(t0_2,t1_2)

p_b0_2 = 2*(1 - stats.t.cdf(np.abs(t0_2), n-p))
print(p_b0_2)

p_b1_2 = 2*(1 - stats.t.cdf(np.abs(t1_2), n-p))
print(p_b1_2)

Regresión de periódico
β0: [12.35140707]
β1: [0.0546931]
0.016575721876358165
0.6214201876094345
[12.35140707] [0.0546931]
(array([0.02154165]), array([0.08784454]))
(array([11.10856669]), array([13.59424744]))
[19.87609562] [3.29959074]
[0.]
[0.0011482]


In [310]:
print("Regresión de radio")
x3 = data["radio"].values.reshape(-1, 1)
y3 = data["sales"].values.reshape(-1, 1)

# Coeficientes
lr3 = LinearRegression().fit(x3, y3)
print("b0_3:", lr3.intercept_)
print("b1_3:", lr3.coef_[0])

# Error estándar de coeficientes
y_pred3 = lr3.predict(x3)
rss3 = np.sum((y3 - y_pred3)**2)
n3 = len(y3)
p3 = 2
sigma2_3 = rss3 / (n3 - p3)

x3_mean = np.mean(x3)
SE3_b1 = np.sqrt(sigma2_3 / np.sum((x3 - x3_mean)**2))
print(SE3_b1)

SE3_b0 = np.sqrt(sigma2_3 * (1/n3 + x3_mean**2 / np.sum((x3 - x3_mean)**2)))
print(SE3_b0)

# Intervalos de confianza (aprox 95% con ±2*SE)
b0_3 = lr3.intercept_
b1_3 = lr3.coef_[0]
print(b0_3, b1_3)

IC_b1_3 = (b1_3 - 2*SE3_b1, b1_3 + 2*SE3_b1)
print(IC_b1_3)

IC_b0_3 = (b0_3 - 2*SE3_b0, b0_3 + 2*SE3_b0)
print(IC_b0_3)

# Estadísticos t y p-values (bilateral)
t0_3 = (b0_3 - 0)/SE3_b0
t1_3 = (b1_3 - 0)/SE3_b1
print(t0_3, t1_3)

p_b0_3 = 2*(1 - stats.t.cdf(np.abs(t0_3), n3 - p3))
print(p_b0_3)

p_b1_3 = 2*(1 - stats.t.cdf(np.abs(t1_3), n3 - p3))
print(p_b1_3)

Regresión de radio
b0_3: [9.3116381]
b1_3: [0.20249578]
0.020411306360090635
0.5629004962718711
[9.3116381] [0.20249578]
(array([0.16167317]), array([0.2433184]))
(array([8.1858371]), array([10.43743909]))
[16.54224531] [9.92076547]
[0.]
[0.]


In [316]:
res = pd.DataFrame({
'medio': ['Periódico (#2)', 'Radio (#3)'],
'b0': [b0_2, b0_3],
'b1': [b1_2, b1_3],
'SE_b1': [SE2_b1, SE3_b1],
't_b1': [t1_2, t1_3],
'p_b1': [p_b1_2, p_b1_3],
'IC95% b1 (inf,sup)': [IC_b1_2, IC_b1_3]
})

print(res.to_string(index=False))

         medio                  b0                    b1    SE_b1                 t_b1                    p_b1                              IC95% b1 (inf,sup)
Periódico (#2) [12.35140706927816] [0.05469309847227336] 0.016576 [3.2995907436334186] [0.0011481958688881466] ([0.021541654719557032], [0.08784454222498969])
    Radio (#3) [9.311638095158283] [0.20249578339243965] 0.020411  [9.920765472824959]                   [0.0]    ([0.1616731706722584], [0.2433183961126209])


## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [255]:
import statsmodels.api as sm

In [275]:
X_tv = data["TV"].values.reshape(-1, 1)

In [283]:
X_radio = data["radio"].values.reshape(-1, 1)

In [291]:
unos = np.ones([n,1])
TV_radio = data["TV"] * data["radio"]

In [293]:
X = data[["TV", "radio", "TV_radio"]]

In [295]:
X = np.column_stack((unos, X_tv,X_radio,TV_radio))

In [297]:
Y = data["sales"]

In [299]:
ols = sm.OLS(Y, X)

In [301]:
results = ols.fit()

In [303]:
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.968
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     1963.
Date:                Thu, 28 Aug 2025   Prob (F-statistic):          6.68e-146
Time:                        17:46:45   Log-Likelihood:                -270.14
No. Observations:                 200   AIC:                             548.3
Df Residuals:                     196   BIC:                             561.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7502      0.248     27.233      0.000       6.261       7.239
x1             0.0191      0.002     12.699      0.000       0.016       0.022
x2             0.0289      0.009      3.241      0.001       0.011       0.046
x3             0.0011   5.24e-05     20.727      0.000       0.001       0.001
==============================================================================
Omnibus:                      128.132   Durbin-Watson:                   2.224
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1183.719
Skew:                          -2.323   Prob(JB):                    9.09e-258
Kurtosis:                      13.975   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

En regresiones simples (sales a TV, sales a radio, sales a newspaper), el p‑value de la pendiente mide la relación total de ese medio con las ventas. En la regresión múltiple (sales a TV + radio + TV×radio), el p‑value mide el efecto parcial de cada variable, es decir, manteniendo las otras fijas. Con los resultados (R² ≈ 0.968), TV, radio y la interacción tienen p‑values muy pequeños y son significativos.

Cambian los p-values entre simple y múltiple porque un medio puede “llevarse crédito” de otro si están relacionados; en múltiple probamos solo su aporte único. Si los predictores se parecen mucho, suben los errores estándar y los p‑values pueden crecer. El aviso de condition number ≈ 1.8e4 sugiere este problema. En el caso de la regresión múltiple, cambia el significado de los efectos principales (p. ej., el efecto de TV cuando radio = 0) y puede mover los p‑values respecto a los modelos simples.

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*